File: Functions/plotCapacityAndResistanceTrending.py
"""
Plot Capacity and Resistance Trending

This function creates trending plots for battery capacity degradation
and resistance growth over time.
Capacity plot shows FEC as text labels at each point.
Resistance plot includes FEC values below date ticks.

Author: Róbinson
Created: November 25, 2025
"""

import matplotlib.pyplot as plt
import numpy as np


def plot_capacity_and_resistance_trending(checkup_capacity_timestamp, checkup_capacity,
                                          checkup_capacity_fec,
                                          checkup_resistance_timestamp, checkup_resistance,
                                          checkup_resistance_fec,
                                          cell_num, cell_label=''):
    """
    Plot capacity and resistance trending over time.
    
    Parameters:
    -----------
    checkup_capacity_timestamp : array-like
        Timestamps for capacity measurements
    checkup_capacity : array-like
        Discharge capacity values in Ah
    checkup_capacity_fec : array-like
        FEC values at capacity measurements
    checkup_resistance_timestamp : array-like
        Timestamps for resistance measurements
    checkup_resistance : array-like
        Resistance values in Ω
    checkup_resistance_fec : array-like
        FEC values at resistance measurements
    cell_num : str
        Cell identifier for plot titles
    cell_label : str, optional
        Descriptive label from test plan (default: '')
    """
    print('\nGenerating capacity and resistance trending plots...')
    
    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(10, 8))
    
    # Capacity degradation over time
    ax1.scatter(checkup_capacity_timestamp, checkup_capacity)
    ax1.set_ylim([20, 62])
    ax1.set_ylabel('Discharge Capacity [Ah]')
    if cell_label:
        ax1.set_title(f'Capacity Degradation Trending {cell_num} - {cell_label}')
    else:
        ax1.set_title(f'Capacity Degradation Trending {cell_num}')
    ax1.grid(True)
    
    # Add FEC values as text below each date tick label
    y_limits1 = ax1.get_ylim()
    y_offset1 = y_limits1[0] - 0.05 * (y_limits1[1] - y_limits1[0])
    for i, (ts, fec) in enumerate(zip(checkup_capacity_timestamp, checkup_capacity_fec)):
        ax1.text(ts, y_offset1, f'({int(round(fec))})', 
                ha='center', va='top', fontsize=8, clip_on=False)
    
    # Create two-line xlabel
    ax1.set_xlabel('Date\\n(FEC)')
    
    # Resistance growth over time
    ax2.scatter(checkup_resistance_timestamp, checkup_resistance)
    ax2.set_ylim([0.5e-3, 10e-3])
    ax2.set_ylabel('Resistance [Ω]')
    if cell_label:
        ax2.set_title(f'Resistance Growth (30s Pulse Method) {cell_num} - {cell_label}')
    else:
        ax2.set_title(f'Resistance Growth (30s Pulse Method) {cell_num}')
    ax2.grid(True)
    
    # Add FEC values as text below each date tick label
    y_limits2 = ax2.get_ylim()
    y_offset2 = y_limits2[0] - 0.05 * (y_limits2[1] - y_limits2[0])
    for i, (ts, fec) in enumerate(zip(checkup_resistance_timestamp, checkup_resistance_fec)):
        ax2.text(ts, y_offset2, f'({int(round(fec))})', 
                ha='center', va='top', fontsize=8, clip_on=False)
    
    # Create two-line xlabel
    ax2.set_xlabel('Date\\n(FEC)')
    
    plt.tight_layout()
    plt.show(block=False)
    plt.pause(0.001)  # Brief pause to ensure plot renders
    
    print('Trending plots created.')

File: Functions/plotCapacityAndResistanceTrending.m
function plotCapacityAndResistanceTrending(checkupCapacityTimeStamp, checkupCapacity_Ah, checkupCapacityFEC, checkupResistanceTimeStamp, checkupResistance_Ohm, checkupResistanceFEC, cellNum, cellLabel)
% plotCapacityAndResistanceTrending - Plot capacity degradation and resistance growth
%
% This function creates a two-panel figure showing capacity fade and
% resistance growth over time from battery checkup measurements.
% Capacity plot shows FEC as text labels at each point.
% Resistance plot includes dual x-axes showing Date and FEC.
%
% Inputs:
%   checkupCapacityTimeStamp    - Datetime array of capacity measurement timestamps
%   checkupCapacity_Ah          - Array of measured discharge capacities (Ah)
%   checkupCapacityFEC          - Array of FEC values at capacity measurements
%   checkupResistanceTimeStamp  - Datetime array of resistance measurement timestamps
%   checkupResistance_Ohm       - Array of measured DC resistance values (Ω)
%   checkupResistanceFEC        - Array of FEC values at resistance measurements
%   cellNum                     - Cell identifier string for plot title
%   cellLabel                   - (Optional) Descriptive label from test plan

if nargin < 8
    cellLabel = '';
end

fprintf('\nGenerating capacity and resistance trending plots...\n');

%%
figure('Units','normalized', 'OuterPosition',[0 0 0.5 1]); % Create a figure that takes up half the screen width

checkupCapacityTimeStamp.Format = 'yyyy-MM-dd'; % don't show the hour

% Use subplot instead of tiledlayout to allow position manipulation for dual x-axis

% Capacity degradation over time (top panel)
x1 = subplot(2,1,1);
scatter(x1, checkupCapacityTimeStamp, checkupCapacity_Ah)
set(x1,'XLim',[min(checkupCapacityTimeStamp) max(checkupCapacityTimeStamp)]);
x1.XTick = checkupCapacityTimeStamp;
x1.XTickLabel = string(checkupCapacityTimeStamp);
ylim([20 62])
ylabel('Discharge Capacity [Ah]');
if isempty(cellLabel)
    title(['Capacity Degradation Trending ', cellNum], 'interpreter', 'none');
else
    title(['Capacity Degradation Trending ', cellNum, ' - ', cellLabel], 'interpreter', 'none');
end
grid on

% Add FEC values as text below each date tick label
yLimits1 = ylim(x1);
yOffset1 = yLimits1(1) - 0.05 * (yLimits1(2) - yLimits1(1));  % Position below axis
hold(x1, 'on');
for i = 1:length(checkupCapacityTimeStamp)
    text(x1, checkupCapacityTimeStamp(i), yOffset1, sprintf('(%d)', round(checkupCapacityFEC(i))), ...
        'HorizontalAlignment', 'center', 'VerticalAlignment', 'top', 'FontSize', 8, 'Clipping', 'off');
end
hold(x1, 'off');

% Create two-line xlabel and move it down to avoid overlap with FEC labels
xlh1 = xlabel(x1, sprintf('Date\n(FEC)'));
xlh1.Position(2) = xlh1.Position(2) - 0.05 * (yLimits1(2) - yLimits1(1));

% Resistance growth over time (bottom panel)

selectedResistanceTime = checkupResistanceTimeStamp;
selectedResistanceTime.Format = 'yyyy-MM-dd';
selectedResistance = checkupResistance_Ohm;

x3 = subplot(2,1,2);
scatter(x3, selectedResistanceTime, selectedResistance);
ylim([0.5e-3 10e-3])
x3.XTick = selectedResistanceTime;
x3.XTickLabel = string(selectedResistanceTime);
xlim([min(selectedResistanceTime) max(selectedResistanceTime)]);
ylabel('Resistance [Ω]');

% Add FEC values as text below each date tick label
yLimits = ylim(x3);
yOffset = yLimits(1) - 0.05 * (yLimits(2) - yLimits(1));  % Position below axis
hold(x3, 'on');
for i = 1:length(selectedResistanceTime)
    text(x3, selectedResistanceTime(i), yOffset, sprintf('(%d)', round(checkupResistanceFEC(i))), ...
        'HorizontalAlignment', 'center', 'VerticalAlignment', 'top', 'FontSize', 8, 'Clipping', 'off');
end
hold(x3, 'off');

% Create two-line xlabel and move it down to avoid overlap with FEC labels
xlh = xlabel(sprintf('Date\n(FEC)'));
xlh.Position(2) = xlh.Position(2) - 0.05 * (yLimits(2) - yLimits(1));
if isempty(cellLabel)
    title(['Resistance Growth (30s Pulse Method) ', cellNum], 'interpreter', 'none');
else
    title(['Resistance Growth (30s Pulse Method) ', cellNum, ' - ', cellLabel], 'interpreter', 'none');
end
grid on

fprintf('Trending plots created. (Elapsed: %.2f s)\n', toc);


if ~isempty(selectedResistanceTime)
    linkaxes([x1 x3],'x')
    MinTime = min(min(checkupCapacityTimeStamp),min(selectedResistanceTime))-minutes(1);
    MaxTime = max(max(checkupCapacityTimeStamp),max(selectedResistanceTime))+minutes(1);
    xlim(x1, [MinTime MaxTime]);  % Explicitly set on x1 (datetime axis)
end
%%
end